In [2]:
import pandas as pd
import numpy as np
import re

In [3]:
df = pd.read_csv('./Datasets/Data_Combinada.csv')

C:\Users\Usuario\AppData\Local\Temp\ipykernel_18060\2478976544.py:1: DtypeWarning: Columns (15,16) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('./Datasets/Data_Combinada.csv')


In [4]:
df.head()

num_expediente       fecha     hora  \
0    2022S000001  01/01/2022  1:30:00   
1    2022S000001  01/01/2022  1:30:00   
2    2022S000002  01/01/2022  0:30:00   
3    2022S000002  01/01/2022  0:30:00   
4    2022S000002  01/01/2022  0:30:00   

                                 localizacion numero  cod_distrito  \
0                          AVDA. ALBUFERA, 19     19          13.0   
1                          AVDA. ALBUFERA, 19     19          13.0   
2  PLAZA. CANOVAS DEL CASTILLO / PASEO. PRADO      2           3.0   
3  PLAZA. CANOVAS DEL CASTILLO / PASEO. PRADO      2           3.0   
4  PLAZA. CANOVAS DEL CASTILLO / PASEO. PRADO      2           3.0   

             distrito           tipo_accidente estado_meteorológico  \
0  PUENTE DE VALLECAS                  Alcance            Despejado   
1  PUENTE DE VALLECAS                  Alcance            Despejado   
2              RETIRO  Colisión fronto-lateral                  NaN   
3              RETIRO  Colisión fronto-lateral                  NaN   
4              RETIRO  Colisión fronto-lateral                  NaN   

             tipo_vehiculo  ...       rango_edad sexo cod_lesividad  \
0                  Turismo  ...  De 30 a 34 años    M           NaN   
1                  Turismo  ...  De 45 a 49 años    H           NaN   
2  Motocicleta hasta 125cc  ...  De 30 a 34 años    H           NaN   
3  Motocicleta hasta 125cc  ...  De 35 a 39 años    M           NaN   
4                  Turismo  ...  De 40 a 44 años    H           NaN   

   lesividad coordenada_x_utm coordenada_y_utm positiva_alcohol  \
0        NaN       443359,226      4472082,272              0.0   
1        NaN       443359,226      4472082,272              0.0   
2        NaN       441155,351      4474129,588              1.0   
3        NaN       441155,351      4474129,588              0.0   
4        NaN       441155,351      4474129,588              0.0   

   positiva_droga  numero_pasajeros  fugado  
0             0.0                 2      No  
1             0.0                 2      No  
2             0.0                 3      No  
3             0.0                 3      No  
4             0.0                 3      No  

[5 rows x 21 columns]

In [5]:
df_vallecas = df[df['distrito'] == "PUENTE DE VALLECAS"].copy()
df_vallecas.head()

num_expediente       fecha     hora         localizacion numero  \
0     2022S000001  01/01/2022  1:30:00   AVDA. ALBUFERA, 19     19   
1     2022S000001  01/01/2022  1:30:00   AVDA. ALBUFERA, 19     19   
30    2022S000015  01/01/2022  7:25:00  AVDA. ENTREVIAS, 94     94   
31    2022S000015  01/01/2022  7:25:00  AVDA. ENTREVIAS, 94     94   
32    2022S000015  01/01/2022  7:25:00  AVDA. ENTREVIAS, 94     94   

    cod_distrito            distrito tipo_accidente estado_meteorológico  \
0           13.0  PUENTE DE VALLECAS        Alcance            Despejado   
1           13.0  PUENTE DE VALLECAS        Alcance            Despejado   
30          13.0  PUENTE DE VALLECAS        Alcance            Despejado   
31          13.0  PUENTE DE VALLECAS        Alcance            Despejado   
32          13.0  PUENTE DE VALLECAS        Alcance            Despejado   

   tipo_vehiculo  ...       rango_edad sexo cod_lesividad  \
0        Turismo  ...  De 30 a 34 años    M           NaN   
1        Turismo  ...  De 45 a 49 años    H           NaN   
30       Turismo  ...  De 35 a 39 años    H           7.0   
31       Turismo  ...  De 40 a 44 años    H          14.0   
32       Turismo  ...  De 35 a 39 años    H           7.0   

                                            lesividad coordenada_x_utm  \
0                                                 NaN       443359,226   
1                                                 NaN       443359,226   
30  Asistencia sanitaria sólo en el lugar del acci...       443634,572   
31                           Sin asistencia sanitaria       443634,572   
32  Asistencia sanitaria sólo en el lugar del acci...       443634,572   

   coordenada_y_utm positiva_alcohol  positiva_droga  numero_pasajeros  fugado  
0       4472082,272              0.0             0.0                 2      No  
1       4472082,272              0.0             0.0                 2      No  
30      4469942,483              0.0             0.0                 3      No  
31      4469942,483              1.0             0.0                 3      No  
32      4469942,483              0.0             0.0                 3      No  

[5 rows x 21 columns]

In [6]:
def strip_num(string):
    return re.sub(r'[\d,]+', '', string)

In [7]:
df_vallecas['localizacion'] = df_vallecas['localizacion'].apply(strip_num)

In [8]:
df_vallecas.head()

num_expediente       fecha     hora      localizacion numero  cod_distrito  \
0     2022S000001  01/01/2022  1:30:00   AVDA. ALBUFERA      19          13.0   
1     2022S000001  01/01/2022  1:30:00   AVDA. ALBUFERA      19          13.0   
30    2022S000015  01/01/2022  7:25:00  AVDA. ENTREVIAS      94          13.0   
31    2022S000015  01/01/2022  7:25:00  AVDA. ENTREVIAS      94          13.0   
32    2022S000015  01/01/2022  7:25:00  AVDA. ENTREVIAS      94          13.0   

              distrito tipo_accidente estado_meteorológico tipo_vehiculo  ...  \
0   PUENTE DE VALLECAS        Alcance            Despejado       Turismo  ...   
1   PUENTE DE VALLECAS        Alcance            Despejado       Turismo  ...   
30  PUENTE DE VALLECAS        Alcance            Despejado       Turismo  ...   
31  PUENTE DE VALLECAS        Alcance            Despejado       Turismo  ...   
32  PUENTE DE VALLECAS        Alcance            Despejado       Turismo  ...   

         rango_edad sexo cod_lesividad  \
0   De 30 a 34 años    M           NaN   
1   De 45 a 49 años    H           NaN   
30  De 35 a 39 años    H           7.0   
31  De 40 a 44 años    H          14.0   
32  De 35 a 39 años    H           7.0   

                                            lesividad coordenada_x_utm  \
0                                                 NaN       443359,226   
1                                                 NaN       443359,226   
30  Asistencia sanitaria sólo en el lugar del acci...       443634,572   
31                           Sin asistencia sanitaria       443634,572   
32  Asistencia sanitaria sólo en el lugar del acci...       443634,572   

   coordenada_y_utm positiva_alcohol  positiva_droga  numero_pasajeros  fugado  
0       4472082,272              0.0             0.0                 2      No  
1       4472082,272              0.0             0.0                 2      No  
30      4469942,483              0.0             0.0                 3      No  
31      4469942,483              1.0             0.0                 3      No  
32      4469942,483              0.0             0.0                 3      No  

[5 rows x 21 columns]

In [9]:
(df_vallecas.groupby('localizacion').count().sort_values(by='num_expediente', ascending=False))

num_expediente  fecha  hora  \
localizacion                                                              
AVDA. ALBUFERA                                         317    317   317   
AUTOV. M- +E                                           123    123   123   
AVDA. ENTREVIAS                                        115    115   115   
AVDA. PABLO NERUDA                                      88     88    88   
AVDA. ALBUFERA / AUTOV. M-                              82     82    82   
...                                                    ...    ...   ...   
AVDA. PABLO NERUDA G                                     1      1     1   
CALL. SIERRA DE ALQUIFE / PUERTO DE ARLA                 1      1     1   
AVDA. PABLO NERUDA / CALL. LEONESES                      1      1     1   
CALL. SIERRA DE CADI / AVDA. ALBUFERA                    1      1     1   
CALL. DOCTOR LOZANO / CALL. PICO CUADRAMON               1      1     1   

                                            numero  cod_distrito  distrito  \
localizacion                                                                 
AVDA. ALBUFERA                                 317           317       317   
AUTOV. M- +E                                   123           123       123   
AVDA. ENTREVIAS                                115           115       115   
AVDA. PABLO NERUDA                              88            88        88   
AVDA. ALBUFERA / AUTOV. M-                      82            82        82   
...                                            ...           ...       ...   
AVDA. PABLO NERUDA G                             1             1         1   
CALL. SIERRA DE ALQUIFE / PUERTO DE ARLA         1             1         1   
AVDA. PABLO NERUDA / CALL. LEONESES              1             1         1   
CALL. SIERRA DE CADI / AVDA. ALBUFERA            1             1         1   
CALL. DOCTOR LOZANO / CALL. PICO CUADRAMON       1             1         1   

                                            tipo_accidente  \
localizacion                                                 
AVDA. ALBUFERA                                         317   
AUTOV. M- +E                                           123   
AVDA. ENTREVIAS                                        115   
AVDA. PABLO NERUDA                                      88   
AVDA. ALBUFERA / AUTOV. M-                              82   
...                                                    ...   
AVDA. PABLO NERUDA G                                     1   
CALL. SIERRA DE ALQUIFE / PUERTO DE ARLA                 1   
AVDA. PABLO NERUDA / CALL. LEONESES                      1   
CALL. SIERRA DE CADI / AVDA. ALBUFERA                    1   
CALL. DOCTOR LOZANO / CALL. PICO CUADRAMON               1   

                                            estado_meteorológico  \
localizacion                                                       
AVDA. ALBUFERA                                               272   
AUTOV. M- +E                                                 109   
AVDA. ENTREVIAS                                              101   
AVDA. PABLO NERUDA                                            75   
AVDA. ALBUFERA / AUTOV. M-                                    75   
...                                                          ...   
AVDA. PABLO NERUDA G                                           0   
CALL. SIERRA DE ALQUIFE / PUERTO DE ARLA                       1   
AVDA. PABLO NERUDA / CALL. LEONESES                            0   
CALL. SIERRA DE CADI / AVDA. ALBUFERA                          0   
CALL. DOCTOR LOZANO / CALL. PICO CUADRAMON                     1   

                                            tipo_vehiculo  tipo_persona  \
localizacion                                                              
AVDA. ALBUFERA                                        315           317   
AUTOV. M- +E                                          122           123   
AVDA. ENTREVIAS                                       115    

Miramos los accidentes en el cuartil más alto del distrito con más accidentes

In [10]:
df_grouped = df_vallecas.groupby('localizacion').count()
counts = df_grouped['num_expediente']

Q1 = counts.quantile(0.25)
Q3 = counts.quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

outliers = counts[(counts < lower_bound) | (counts > upper_bound)]

percentile_95 = counts.quantile(0.95)

top_5_percent_df = df_grouped[df_grouped['num_expediente'] >= percentile_95].reset_index()

top_5_percent_df = top_5_percent_df.sort_values(by=['num_expediente'], ascending=False)

Vamos a ver donde se concentran los accidentes, mirando ahora la imagen completa, con el número de calle, y los tipos de accidentes

In [11]:
filtered_df_vallecas = df_vallecas[df_vallecas['localizacion'].isin(top_5_percent_df['localizacion'])].copy()
filtered_df_vallecas.sort_values(by=['localizacion'])

num_expediente       fecha      hora        localizacion   numero  \
26493    2022S022610  02/08/2022  18:10:00        AUTOV. A- +E  +00400E   
13166    2022S011020  20/04/2022  18:15:00        AUTOV. A- +E  +00500E   
31734    2022S027913  26/09/2022   7:00:00        AUTOV. A- +E  +00500E   
31733    2022S027913  26/09/2022   7:00:00        AUTOV. A- +E  +00500E   
31732    2022S027913  26/09/2022   7:00:00        AUTOV. A- +E  +00500E   
...              ...         ...       ...                 ...      ...   
39301    2022S035733  21/11/2022   9:15:00  CMNO. VALDERRIBAS       120   
49243    2023S003225  03/02/2023   8:15:00  CMNO. VALDERRIBAS       113   
58049    2023S015043  12/04/2023  20:30:00  CMNO. VALDERRIBAS        99   
58047    2023S015043  12/04/2023  20:30:00  CMNO. VALDERRIBAS        99   
73453    2023S027263  12/08/2023   1:25:00  CMNO. VALDERRIBAS       113   

       cod_distrito            distrito                tipo_accidente  \
26493          13.0  PUENTE DE VALLECAS             Colisión múltiple   
13166          13.0  PUENTE DE VALLECAS                       Alcance   
31734          13.0  PUENTE DE VALLECAS                       Alcance   
31733          13.0  PUENTE DE VALLECAS                       Alcance   
31732          13.0  PUENTE DE VALLECAS                       Alcance   
...             ...                 ...                           ...   
39301          13.0  PUENTE DE VALLECAS              Colisión lateral   
49243          13.0  PUENTE DE VALLECAS                       Alcance   
58049          13.0  PUENTE DE VALLECAS  Choque contra obstáculo fijo   
58047          13.0  PUENTE DE VALLECAS  Choque contra obstáculo fijo   
73453          13.0  PUENTE DE VALLECAS  Choque contra obstáculo fijo   

      estado_meteorológico  tipo_vehiculo  ...       rango_edad sexo  \
26493                  NaN        Turismo  ...  De 45 a 49 años    H   
13166            Despejado        Turismo  ...  De 30 a 34 años    H   
31734            Despejado        Turismo  ...  De 45 a 49 años    M   
31733            Despejado        Turismo  ...  De 50 a 54 años    H   
31732            Despejado  Camión rígido  ...  De 40 a 44 años    H   
...                    ...            ...  ...              ...  ...   
39301              Nublado        Turismo  ...  De 30 a 34 años    H   
49243                  NaN        Turismo  ...  De 40 a 44 años    M   
58049              Nublado      Furgoneta  ...  De 55 a 59 años    H   
58047              Nublado      Furgoneta  ...  De 55 a 59 años    H   
73453            Despejado        Turismo  ...      Desconocido    H   

      cod_lesividad                                          lesividad  \
26493           NaN                                                NaN   
13166          14.0                           Sin asistencia sanitaria   
31734           NaN                                                NaN   
31733           NaN                                                NaN   
31732           NaN                                                NaN   
...             ...                                                ...   
39301           NaN                                                NaN   
49243           NaN                                                NaN   
58049          14.0                           Sin asistencia sanitaria   
58047           7.0  Asistencia sanitaria sólo en el lugar del acci...   
73453           NaN                                                NaN   

      coordenada_x_utm coordenada_y_utm positiva_alcohol  positiva_droga  \
26493       441976,313      4469760,365              0.0             0.0   
13166       442028,358      4469677,005              0.0             0.0   
31734       442086,031      4469589,003              0.0             0.0   
31733       442086,031      4469589,003              0.0             0.0   
31732       442086,031      4469589,003              0.0             0.0   
...               

In [12]:
filtered_df_vallecas['coordenada_x_utm'] = pd.to_numeric(filtered_df_vallecas['coordenada_x_utm'].str.replace(',', '.'), errors='coerce')
filtered_df_vallecas['coordenada_y_utm'] = pd.to_numeric(filtered_df_vallecas['coordenada_y_utm'].str.replace(',', '.'), errors='coerce')

filtered_df_vallecas.dropna(subset=['coordenada_x_utm', 'coordenada_y_utm'], inplace=True)

In [13]:
filtered_df_vallecas

num_expediente       fecha      hora                localizacion numero  \
0        2022S000001  01/01/2022   1:30:00             AVDA. ALBUFERA      19   
1        2022S000001  01/01/2022   1:30:00             AVDA. ALBUFERA      19   
30       2022S000015  01/01/2022   7:25:00            AVDA. ENTREVIAS      94   
31       2022S000015  01/01/2022   7:25:00            AVDA. ENTREVIAS      94   
32       2022S000015  01/01/2022   7:25:00            AVDA. ENTREVIAS      94   
...              ...         ...       ...                         ...    ...   
65455    2023S022129  05/06/2023  20:15:00  AVDA. ALBUFERA / AUTOV. M-    302   
65456    2023S022129  05/06/2023  20:15:00  AVDA. ALBUFERA / AUTOV. M-    302   
65512    2023S022164  06/06/2023  11:25:00         AVDA. PABLO NERUDA      50   
65513    2023S022164  06/06/2023  11:25:00         AVDA. PABLO NERUDA      50   
65514    2023S022164  06/06/2023  11:25:00         AVDA. PABLO NERUDA      50   

       cod_distrito            distrito     tipo_accidente  \
0              13.0  PUENTE DE VALLECAS            Alcance   
1              13.0  PUENTE DE VALLECAS            Alcance   
30             13.0  PUENTE DE VALLECAS            Alcance   
31             13.0  PUENTE DE VALLECAS            Alcance   
32             13.0  PUENTE DE VALLECAS            Alcance   
...             ...                 ...                ...   
65455          13.0  PUENTE DE VALLECAS   Colisión lateral   
65456          13.0  PUENTE DE VALLECAS   Colisión lateral   
65512          13.0  PUENTE DE VALLECAS  Colisión múltiple   
65513          13.0  PUENTE DE VALLECAS  Colisión múltiple   
65514          13.0  PUENTE DE VALLECAS  Colisión múltiple   

      estado_meteorológico tipo_vehiculo  ...       rango_edad sexo  \
0                Despejado       Turismo  ...  De 30 a 34 años    M   
1                Despejado       Turismo  ...  De 45 a 49 años    H   
30               Despejado       Turismo  ...  De 35 a 39 años    H   
31               Despejado       Turismo  ...  De 40 a 44 años    H   
32               Despejado       Turismo  ...  De 35 a 39 años    H   
...                    ...           ...  ...              ...  ...   
65455              Nublado       Turismo  ...  De 25 a 29 años    H   
65456              Nublado       Turismo  ...  De 35 a 39 años    H   
65512                  NaN     Furgoneta  ...  De 40 a 44 años    H   
65513                  NaN       Turismo  ...      Desconocido    H   
65514                  NaN       Turismo  ...      Desconocido    H   

      cod_lesividad                                          lesividad  \
0               NaN                                                NaN   
1               NaN                                                NaN   
30              7.0  Asistencia sanitaria sólo en el lugar del acci...   
31             14.0                           Sin asistencia sanitaria   
32              7.0  Asistencia sanitaria sólo en el lugar del acci...   
...             ...                                                ...   
65455           NaN                                                NaN   
65456           NaN                                                NaN   
65512           NaN                                                NaN   
65513           NaN                                                NaN   
65514           NaN                                                NaN   

      coordenada_x_utm  coordenada_y_utm  positiva_alcohol  positiva_droga  \
0           443359.226       4472082.272               0.0             0.0   
1           443359.226       4472082.272               0.0             0.0   
30          443634.572       4469942.483               0.0             0.0   
31          443634.572       4469942.483               1.0             0.0   
32          443634.572       4469942.483               0.0             0.0   
...                ...               ...               ...             ...   


In [14]:
import utm
import folium

def utm_to_latlon(utm_x, utm_y, zone_number=30, zone_letter='T'):
    """
    Convert UTM coordinates to Latitude and Longitude.
    Assumes zone number 30 and zone letter 'T' for Madrid.
    """
    try:
        lat, lon = utm.to_latlon(utm_x, utm_y, zone_number, zone_letter)
        return lat, lon
    except Exception as e:
        print("Error en la conversión:", e, " utmx: ", utm_x, "utmy: ", utm_y, " ")
        return None, None

# Processing the data: converting UTM coordinates to latitude and longitude
coordinates = filtered_df_vallecas.filter(items=['coordenada_x_utm', 'coordenada_y_utm', 'tipo_accidente', 'tipo_vehiculo'])
coordinates.columns = ['utm_x', 'utm_y', 'tipo_accidente', 'tipo_vehiculo']
# Apply the conversion
coordinates['lat'], coordinates['lon'] = zip(*coordinates.apply(lambda row: utm_to_latlon(row['utm_x'], row['utm_y']), axis=1))
# Drop rows with NaN values in lat/lon
coordinates.dropna(subset=['lat', 'lon'], inplace=True)

# Creating a map centered around Madrid
madrid_map = folium.Map(location=[40.4168, -3.7038], zoom_start=12)

# Adding the coordinates to the map
for idx, row in coordinates.iterrows():
    label = f'Lat: {row["lat"]}, Lon: {row["lon"]}, \nAccidente: {row["tipo_accidente"]}, \nVehículo: {row["tipo_vehiculo"]}'
    folium.Marker([row['lat'], row['lon']], popup=label).add_to(madrid_map)
# Save the map to an HTML file
map_file = 'Maps/madrid_map.html'
madrid_map.save(map_file)

map_file

'Maps/madrid_map.html'

Vamos a filtrar todas las carreteras

In [15]:
def path_denomination(path:str):
    return path.split()[0].upper().replace(',', '').replace('.', '')

In [16]:
df['first_word'] = df['localizacion'].apply(path_denomination)
result = df.groupby('first_word')['localizacion'].count()
result.sort_values(ascending=False)

first_word
CALL        41024
AVDA        12934
AUTOV        7212
PASEO        4690
PLAZA        2501
            ...  
LEONARDO        1
LATINA          1
ENLACE          1
RETIRO          1
ZONA            1
Name: localizacion, Length: 343, dtype: int64

# Estudio Medidas implementadas en la M-30 desde el 2022
A continuación iniciaremos un análisis exploratorio sobre las medidas ZBE implementadas a partir del año 2022 en la ciudad de Madrid:
En resumen las Medidas Implementadas:
1. Zona de Bajas Emisiones (ZBE):

**En enero de 2022**, se prohibió el acceso y la circulación de vehículos sin distintivo ambiental en las vías públicas urbanas del interior de la M-30 (excluyendo la propia M-30).
**En enero de 2023**, se extendió la prohibición al propio tramo de la M-30.
**En enero de 2024**, la prohibición cubrió todas las vías públicas urbanas del municipio de Madrid.
**Para enero de 2025**, se espera que la prohibición se extienda a los turismos A domiciliados en Madrid y a otros vehículos A no turismos (camiones, furgonetas, motocicletas y ciclomotores) que hasta el 31 de diciembre de 2024 no tenían restricciones.

2. Instalación de cámaras y sistemas de control:

Para vigilar y controlar la ZBE, se han instalado 207 cámaras de tráfico adicionales en el exterior de la M-30, sumándose a las 257 cámaras y 37 semáforos con control foto-rojo ya existentes.

In [17]:
# Lee el archivo CSV de 2021
data_2021 = pd.read_csv("./Datasets/2021_Accidentalidad.csv", sep=";")

# Convierte la columna 'fecha' a datetime para poder filtrar por año en ambos datasets
df['fecha'] = pd.to_datetime(df['fecha'], format='%d/%m/%Y')
data_2021['fecha'] = pd.to_datetime(data_2021['fecha'], format='%d/%m/%Y')

# Filtra los datos de 2022 y 2023 a partir del conjunto de datos 'df'
data_2022 = df[df['fecha'].dt.year == 2022]
data_2023 = df[df['fecha'].dt.year == 2023]

# Filtrar los registros específicos de la M-30 para 2021, 2022 y 2023
m30_2021 = data_2021[data_2021['localizacion'].str.contains('M-30')]
m30_2022 = data_2022[data_2022['localizacion'].str.contains('M-30')]
m30_2023 = data_2023[data_2023['localizacion'].str.contains('M-30')]

# Calculando el número de registros mensuales para cada año para la M-30
m30_2021['mes'] = m30_2021['fecha'].dt.month
m30_2022['mes'] = m30_2022['fecha'].dt.month
m30_2023['mes'] = m30_2023['fecha'].dt.month

registros_mensuales_2021 = m30_2021.groupby('mes').size()
registros_mensuales_2022 = m30_2022.groupby('mes').size()
registros_mensuales_2023 = m30_2023.groupby('mes').size()

registros_mensuales_2021, registros_mensuales_2022, registros_mensuales_2023

C:\Users\Usuario\AppData\Local\Temp\ipykernel_18060\2716251663.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  m30_2021['mes'] = m30_2021['fecha'].dt.month
C:\Users\Usuario\AppData\Local\Temp\ipykernel_18060\2716251663.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  m30_2022['mes'] = m30_2022['fecha'].dt.month
C:\Users\Usuario\AppData\Local\Temp\ipykernel_18060\2716251663.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,c

(mes
 1     185
 2     277
 3     359
 4     276
 5     246
 6     281
 7     235
 8     144
 9     512
 10    478
 11    472
 12    454
 dtype: int64,
 mes
 1     276
 2     367
 3     308
 4     347
 5     370
 6     314
 7     259
 8     171
 9     274
 10    289
 11    402
 12    434
 dtype: int64,
 mes
 1    310
 2    330
 3    309
 4    250
 5    299
 6    231
 7    204
 8    175
 dtype: int64)

Hemos realizado los registros mensuales de accidentes en la M-30 para los años 2021, 2022 y 2023 (hasta septiembre). A continuación, se presenta el número de registros por mes para cada año, lo que nos permitirá evaluar cómo ha evolucionado el tráfico en la M-30 en relación con las medidas implementadas.

### Número de Registros por Mes en la M-30:
**2021**:
Enero: 185
Febrero: 277
Marzo: 359
Abril: 276
Mayo: 246
Junio: 281
Julio: 235
Agosto: 144
Septiembre: 512
Octubre: 478
Noviembre: 472
Diciembre: 454

**2022**:
Enero: 276
Febrero: 367
Marzo: 308
Abril: 347
Mayo: 370
Junio: 314
Julio: 259
Agosto: 171
Septiembre: 274
Octubre: 289
Noviembre: 402
Diciembre: 434

**2023 (hasta septiembre)**:
Enero: 310
Febrero: 330
Marzo: 309
Abril: 250
Mayo: 299
Junio: 231
Julio: 204
Agosto: 175

#### Análisis Comparativo:
Al comparar los datos de 2021 y 2022, se observa un incremento en la mayoría de los meses en 2022. Esto podría deberse a una recuperación del tráfico tras las restricciones de la pandemia o a cambios en la detección y reporte de incidentes.
Para 2023, aunque los datos están disponibles solo hasta septiembre, se observa una tendencia a la disminución en la cantidad de registros en comparación con 2022, particularmente en los meses más recientes (junio a agosto). Esto podría ser indicativo de un impacto positivo de las medidas implementadas en la M-30, especialmente considerando que en enero de 2023, la prohibición se extendió al propio tramo de la M-30.

#### **Conclusiones**:
La tendencia decreciente en los últimos meses de 2023 podría ser un indicador temprano de la efectividad de las medidas de ZBE implementadas en la M-30, aunque se necesita más tiempo y datos para confirmar esta tendencia y entender su impacto total. Además, se deberían considerar factores externos que puedan afectar el tráfico y la incidencia de accidentes, como cambios en los patrones de movilidad urbana, el clima, y eventos específicos en la ciudad.

Además, dicha medida no fue del todo implementada para mejorar el tráfico únicamente, sino para mejorar la calidad del aire en la ciudad.
#### Impactos y Observaciones de la Medida ZBE:
**-Reducción de la Contaminación**: En 2023, los niveles de dióxido de nitrógeno en Madrid fueron más bajos que el año anterior, lo que indica una mejora en la calidad del aire de la ciudad, atribuida a la implementación de la ZBE​​.
**-Cumplimiento de Normativas Europeas**: El 2022 fue el primer año en que Madrid cumplió con la directiva europea de calidad del aire, lo que sugiere un impacto positivo de las medidas de tráfico implementadas​​.
**-Reducción de Vehículos Contaminantes**: Se observó una disminución en el número de vehículos más contaminantes (matrícula A) hasta un 2.3% del total en el interior de la ciudad. Las cámaras detectan diariamente unos 2.000 vehículos tipo A sancionables​

## Análisis de tramos con mayor Índice de accidentes de la autovía M-30
A continuación haremos un estudio sobre los tramos con mayor cantidad de accidentes en la autovía M-30 para acercarnos algo a lo que puede ser el causante de este problema

In [23]:
# Filtrando data solo para la autov. M-30
m30_data = df[df['localizacion'].str.contains('M-30', na=False)].copy()

m30_data_head = m30_data.head()

m30_data_head

num_expediente      fecha     hora  \
27    2022S000013 2022-01-01  8:10:00   
35    2022S000018 2022-01-01  9:00:00   
36    2022S000018 2022-01-01  9:00:00   
66    2022S000032 2022-01-02  1:00:00   
81    2022S000049 2022-01-01  7:10:00   

                                    localizacion   numero  cod_distrito  \
27                          AUTOV. M-30, +01100I  +01100I           2.0   
35  AUTOV. M-30, +01100I PTO KM. 11,300 INTERIOR  +01100I           2.0   
36  AUTOV. M-30, +01100I PTO KM. 11,300 INTERIOR  +01100I           2.0   
66                      CALL. VADO / AUTOV. M-30        1          12.0   
81                    AUTOV. M-30, +00900E, 09XC  +00900E          14.0   

      distrito                tipo_accidente estado_meteorológico  \
27  ARGANZUELA  Choque contra obstáculo fijo            Despejado   
35  ARGANZUELA       Colisión fronto-lateral            Despejado   
36  ARGANZUELA       Colisión fronto-lateral            Despejado   
66       USERA  Choque contra obstáculo fijo            Despejado   
81   MORATALAZ                       Alcance            Despejado   

   tipo_vehiculo  ... sexo cod_lesividad  \
27       Turismo  ...    H           NaN   
35       Turismo  ...    H          14.0   
36       Turismo  ...    M           1.0   
66     Furgoneta  ...    H           NaN   
81       Turismo  ...    M           NaN   

                                      lesividad  coordenada_x_utm  \
27                                          NaN        442679,507   
35                     Sin asistencia sanitaria        442366,272   
36  Atención en urgencias sin posterior ingreso        442366,272   
66                                          NaN        440991,275   
81                                          NaN        443429,899   

   coordenada_y_utm positiva_alcohol positiva_droga  numero_pasajeros  fugado  \
27      4471637,799              1.0            0.0                 1      No   
35      4471296,603              0.0            0.0                 2      No   
36      4471296,603              0.0            0.0                 2      No   
66       4470831,56              1.0            0.0                 1      No   
81      4472683,605              0.0            0.0                 6      No   

    first_word  
27       AUTOV  
35       AUTOV  
36       AUTOV  
66        CALL  
81       AUTOV  

[5 rows x 22 columns]

In [24]:
# Agrupando por 'localizacion' y contando
accident_counts = m30_data.groupby('localizacion').size().reset_index(name='accident_count')

top_accident_locations = accident_counts.sort_values(by='accident_count', ascending=False)

top_accident_locations

localizacion  accident_count
405                         AUTOV. M-30, +01100I              63
328                         AUTOV. M-30, +00800I              61
372                         AUTOV. M-30, +01000E              54
381                         AUTOV. M-30, +01000I              41
598                         AUTOV. M-30, +03200E              38
...                                          ...             ...
588   AUTOV. M-30, +03100I, M-30 PK 30.900 Cª 1ª               1
593                        AUTOV. M-30, +031500I               1
1439     AUTOV. M-30, CALZADA INTERIOR PK 31.800               1
594                AUTOV. M-30, +031700 INTERIOR               1
0      ACCESO DE M-23 DE SALIDA A AUTOV. M-30 C2               1

[1834 rows x 2 columns]

Los tramos de la M-30 con el mayor número de accidentes son:

1. AUTOV. M-30, +01100I con 63 accidentes.
2. AUTOV. M-30, +00800I con 61 accidentes.
3. AUTOV. M-30, +01000E con 54 accidentes.
4. AUTOV. M-30, +01000I con 41 accidentes.
5. AUTOV. M-30, +03200E con 38 accidentes.
6. AUTOV. A-2 / AUTOV. M-30 con 34 accidentes.
7. AUTOV. M-30, +00100E con 29 accidentes.
8. AUTOV. M-30, +01200E con 28 accidentes.
9. AUTOV. M-30, +00700I con 28 accidentes.
10. CALL. ALCALA / AUTOV. M-30 con 26 accidentes.

In [25]:
from my_map_creator import MapCreator
m30_data['coordenada_x_utm'] = pd.to_numeric(m30_data['coordenada_x_utm'].str.replace(',', '.'), errors='coerce')
m30_data['coordenada_y_utm'] = pd.to_numeric(m30_data['coordenada_y_utm'].str.replace(',', '.'), errors='coerce')
mapa_M30 = MapCreator(m30_data, lat_col_name= "coordenada_x_utm", long_col_name="coordenada_y_utm")

In [26]:
mapa_M30.create_map(map_name= "mapa_M30")